### ISSUE TITLE
Use components in each blocks?

### ISSUE THREAD

**Issue by samselikoff** [tedconf/code-style-guides#23]
Dockyard's original guide makes this suggestions:

### Use components in `{{#each}}` blocks

Contents of your each blocks should be a single line, use components
when more than one line is needed. This will allow you to test the
contents in isolation via unit tests, as your loop will likely contain
more complex logic in this case.

```hbs
{{! Good }}
{{#each posts as |post|}}
  {{post-summary post=post}}
{{/each}}

{{! Bad }}
{{#each posts as |post|}}
  <article>
    <img src={{post.image}} />
    <h1>{{post.title}}</h2>
    <p>{{post.summar}}</p>
  </article>
{{/each}}
```

---

Do we really want to be this stringent? I personally say, no. What would be the unit test for the chunk of hbs above?

I vote we relax this a bit to a looser suggestion: "The contents of `{{#each}}` blocks have a habit of becoming complicated. When there's too much going on, extract the contents out into a new component" with a simple example.

Thoughts @brenna @ryanto ?

**Comment by brenna** [comment]
What about modifying the line limit to feel more appropriate? 
> Contents of your each blocks should be no more than 5 lines, use components
when more than 5 lines are needed.

I don't want to get too dogmatic about # of lines, since that doesn't always correlate to complexity, but a quantifiable measures is easier to adhere to compared to discretionary terms like "too much going on."  I'd like these style guides to be dead simple to follow for someone new to Ember. Making exceptions to the rules like this will likely happen at an expert level, but setting an explicit rule is a good base.

**Comment by samselikoff** [comment]
I agree, similar to Sandi's 5 rules for rails. They should be a guideline, and you can break them, but only if you have a good reason. And beginners should follow them more strictly, until they understand the tradeoffs involved in breaking them better.

I like your suggestion, I would say maybe even something more, somewhere around 5-10 feels right. Also I find that when I am adding actions that just pertain to elements within the `each`, that's really what blossoms the complexity of the parent context. So maybe our guidelines could be

 - 6+ lines, OR
 - 2+ actions specifically related to what's going on within the block

Thoughts?

**Comment by brenna** [comment]
My instinct says keep it simple and just go with "n lines maximum." Something like action complexity will reveal itself and cause you to split it up or just fall under the "5 lines" rule. But I don't feel too strongly about this, as long as it's a measurable rule.
